### PR TITLE
Add RequiresMountsFor= to systemd generate

### PR DESCRIPTION
**Pull request by tunacado**
It is rare but possible that storage locations for the graphroot and the
runroot are not mounted at boot time, and therefore might race when
doing container operations.  An example we've seen in the wild is that a
slow tmpfs mount for the runroot would suddenly mount over /run, causing
the container to lose all currently-running data, requiring a system
refresh to get it back.

This patch adds RequiresMountsFor= to the systemd.unit header to ensure
the paths for both the graphroot and runroot are mounted prior to
starting any generated unit files.

Signed-off-by: Robb Manes <rmanes@redhat.com>

### CODE REVIEW

**Comment by tunacado** [comment]
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1939010, no GitHub issue for this yet.

Using it as an example:
```bash
# podman run -d --name nginx docker.io/nginx

# podman generate systemd nginx
# container-7abce23d45d38c2121ba8d57b355672d8e4c07fb62b80b9353db795104637417.service
# autogenerated by Podman 3.1.0-dev
# Tue Mar 16 13:07:09 EDT 2021

[Unit]
Description=Podman container-7abce23d45d38c2121ba8d57b355672d8e4c07fb62b80b9353db795104637417.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
After=network-online.target
RequiresMountsFor=/var/lib/containers/storage /var/run/containers/storage

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
Restart=on-failure
TimeoutStopSec=70
ExecStart=/home/rmanes/go/src/github.com/tunacado/podman/bin/podman start 7abce23d45d38c2121ba8d57b355672d8e4c07fb62b80b9353db795104637417
ExecStop=/home/rmanes/go/src/github.com/tunacado/podman/bin/podman stop -t 10 7abce23d45d38c2121ba8d57b355672d8e4c07fb62b80b9353db795104637417
ExecStopPost=/home/rmanes/go/src/github.com/tunacado/podman/bin/podman stop -t 10 7abce23d45d38c2121ba8d57b355672d8e4c07fb62b80b9353db795104637417
PIDFile=/var/run/containers/storage/overlay-containers/7abce23d45d38c2121ba8d57b355672d8e4c07fb62b80b9353db795104637417/userdata/conmon.pid
Type=forking

[Install]
WantedBy=multi-user.target default.target

# podman generate systemd nginx >> /etc/systemd/system/nginx-podman.service

# systemctl daemon-reload

# systemctl reboot

# podman ps -a
CONTAINER ID  IMAGE            COMMAND               CREATED             STATUS             PORTS   NAMES
7abce23d45d3 docker.io/nginx  nginx -g daemon o...  About a minute ago  Up 10 seconds ago          nginx
```

**Comment by tunacado** [comment]
I see I need to add some regression tests to `containers_test.go` here as well.

**Comment by vrothberg** [comment]
Thanks for pulling this together, @tunacado!

The fix looks good to me, but the tests need to me massaged a bit to account for the new field.

@giuseppe PTAL

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by tunacado** [comment]
Working on the tests, will update for further review ASAP.

**Comment by rhatdan** [comment]
@tunacado Thanks, are you still working on this?

I would like to get this in, for podman 3.1.

**Comment by rhatdan** [comment]
@vrothberg PTAL
@tunacado Could you just add a test to generate a unit file and make sure the RequiresMountsFor= line is in the generated unit file and is correct.

**Comment by tunacado** [comment]
> @tunacado Could you just add a test to generate a unit file and make sure the RequiresMountsFor= line is in the generated unit file and is correct.

Yes, sorry about the delay - something unexpected came up.  Local tests pass and both containers and pods generate successfully, I just pushed a new version of the PR now.  Thanks!

**Comment by rhatdan** [comment]
You missed updating the podman-systemd-generate man page.

**Review by vrothberg — COMMENTED** [review]
LGTM, great work

**Comment by tunacado** [comment]
> You missed updating the podman-systemd-generate man page.

I'm not sure what to add; I didn't change any options that a user can interact with.  I can change the examples to include `RequiresMountsFor` as well, but it looks like a lot of the unit headers haven't been updated in the examples section.  Should I just add `RequiresMountsFor` and maybe any of the other missing fields to the man page?  Other than that, I am not sure specifically what in the man page to change other than the examples.

**Comment by vrothberg** [comment]
> > You missed updating the podman-systemd-generate man page.
> 
> I'm not sure what to add; I didn't change any options that a user can interact with. I can change the examples to include `RequiresMountsFor` as well, but it looks like a lot of the unit headers haven't been updated in the examples section. Should I just add `RequiresMountsFor` and maybe any of the other missing fields to the man page? Other than that, I am not sure specifically what in the man page to change other than the examples.

Adding `RequiresMountsFor` to the examples sounds good to me.  A sentence or two about _why_ we need it would be great as well.  My brain is an efficient garbage collector, especially for solved problems :)  Having it in the man page would be a huge help to refresh my memory when needed.

**Comment by tunacado** [comment]
> Adding `RequiresMountsFor` to the examples sounds good to me. A sentence or two about _why_ we need it would be great as well. My brain is an efficient garbage collector, especially for solved problems :) Having it in the man page would be a huge help to refresh my memory when needed.

Gotcha; changes added to where I think they go.  Would appreciate a review.  Thanks!

File: docs/source/markdown/podman-generate-systemd.1.md [hunk 1]
```diff
@@ -61,7 +61,7 @@ Set the systemd unit name separator between the name/id of a container/pod and t
 
 ### Generate and print a systemd unit file for a container
 
-Generate a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout to stdout.
+Generate a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout to stdout. Note that the **IRequiresMountsFor** option in the **Unit** section ensures that the container storage for both the GraphRoot and the RunRoot are mounted prior to starting the service. For systems with container storage on disks like iSCSI or other remote block protocols, this ensures that podman is not executed prior to any necessary storage operations coming online.
```
Comment (vrothberg) [inline docs/source/markdown/podman-generate-systemd.1.md:64]:
```suggestion
Generate a systemd unit file for a container running nginx with an *always* restart policy and 1-second timeout to stdout. Note that the **RequiresMountsFor** option in the **Unit** section ensures that the container storage for both the GraphRoot and the RunRoot are mounted prior to starting the service. For systems with container storage on disks like iSCSI or other remote block protocols, this ensures that Podman is not executed prior to any necessary storage operations coming online.
```

Comment (tunacado) [reply docs/source/markdown/podman-generate-systemd.1.md]: Fixed, thanks!

**Review by vrothberg — COMMENTED** [review]
LGTM, thanks a lot. Great work!

**Comment by mheon** [comment]
LGTM
/hold
/lgtm

**Pull request merged**